### PR TITLE
Member profile page

### LINKED ISSUE #8: Member Profile Page
- Displays all the fields in spreadsheet 
- Make everything input fields (disable if no access)

### PR DESCRIPTION
**Pull request by mattwalo32**
## Summary

Create a profile page allowing members to view and edit information. All of the content on the page is dynamically generated -- that is, if the member schema is changed on the backend, this page will automatically reflect this (as long as it's not a new type)

## Changes

- Create profile page
- Create endpoint to get enum options
- Create endpoint to get schema types
- Connect profile page to backend
- Disable fields that users don't have permission to edit

## Screenshots

image
image

## Requests / Responses 

**Response**
GET `/api/members/options` Returns a list enum options by attribute.

This allows the frontend to dynamically populate the dropdowns.

```
{
classStanding: [ {label: "FRESHMAN", value: "FRESHMAN"},
{label: "SOPHOMORE", value: "SOPHOMORE"},
{label: "JUNIOR", value: "JUNIOR"},
{label: "SPRING_SENIOR", value: "SPRING_SENIOR"},
{label: "FALL_SENIOR", value: "FALL_SENIOR"},
{label: "MASTERS", value: "MASTERS"},
{label: "PHD", value: "PHD"}]

generationSemester: [ {label: "FALL", value: "FALL"},
{label: "SPRING", value: "SPRING"}]
...
}
```

**Response**
GET `/api/members/schema` Returns all member attributes by their types

This allows the frontend to dynamically select the appropriate way to display information.

```
{
Boolean: ["areDuesPaid"]
Date: ["birthdate"]
Enum: ["gradSemester", "classStanding", "generationSemester", "location", "role", "level", "status"]
Number: ["gradYear", "generationYear", "__v"]
ObjectID: ["_id"]
String: ["firstName", "lastName", "oauthID", "email", "phone"]
}
```

Closes #8

### CODE REVIEW

File: client/src/pages/Member.js
Comment (Qrtn) [inline client/src/pages/Member.js:16]: Use `let` instead of `var`

File: api/src/api/members.js
Comment (Qrtn) [inline api/src/api/members.js:89]: Rather than explicitly specifying which fields have options & their corresponding options, I'd suggest
```es6
const options = {};
Member.schema.eachPath((pathname, schemaType) => {
  const { enum: optionsEnum } = schemaType.options;
  if (optionsEnum) {
    options[pathname] = optionsEnum;
  }
});
```
Also `options[level]` would be an array of options now. `options`:
```es6
{
  gradSemester: [ 'FALL', 'SPRING' ],
  classStanding: [
    'FRESHMAN',
    'SOPHOMORE',
    'JUNIOR',
    'SPRING_SENIOR',
    'FALL_SENIOR',
    'MASTERS',
    'PHD'
  ],
  generationSemester: [ 'FALL', 'SPRING' ],
  location: [ 'REMOTE', 'CAMPUS', 'TBD' ],
  role: [
    'CO_DIRECTOR',
    'COMMUNITY_DIRECTOR',
    'EXTERNAL_DIRECTOR',
    'ACADEMY_LEAD',
    'TECH_LEAD',
    'PR_LEAD',
    'PRODUCT_MANAGER',
    'PRODUCT_DESIGNER',
    'DEVELOPER',
    'ACADEMY_MEMBER'
  ],
  level: [ 'ADMIN', 'DIRECTOR', 'LEAD', 'MEMBER', 'TBD' ],
  status: [ 'ACTIVE', 'HIATUS', 'RETIRED', 'TBD' ]
}
```

File: api/src/models/member.js
Comment (Qrtn) [inline api/src/models/member.js:116]: This is already present in Member, you don't have to explicitly export it.

File: api/src/api/members.js
Comment (Qrtn) [inline api/src/api/members.js:95]: I actually need to change this in `validateMemberQuery` too, but I just remembered a better way to loop through object keys, values, and entries. https://zellwk.com/blog/looping-through-js-objects/

So here I'd suggest `for (const [attributeLabel, optionsEnum] of Object.entries(options))`

File: api/src/api/members.js
Comment (Qrtn) [inline api/src/api/members.js:97]: This could be `labeledOptions[attributeLabel] = optionsEnum.map((option) => { label: option, value: option });` and can replace lines 98-108

File: api/src/models/member.js
Comment (Qrtn) [inline api/src/models/member.js:122]: If you dynamically generate the options, you don't have to export these enums.

File: client/src/components/EditableAttribute/BooleanAttribute.js
Comment (Qrtn) [inline client/src/components/EditableAttribute/BooleanAttribute.js:6]: Maybe `const VALUE_OPTIONS` to signify it's an actual constant?

File: client/src/pages/Member.js
Comment (Qrtn) [inline client/src/pages/Member.js:16]: Best to use `let` instead of `var`

**Review by Qrtn — CHANGES_REQUESTED** [review]
Great work Matt, left a few suggestions.

File: api/src/api/members.js
Comment (Qrtn) [reply api/src/api/members.js:97]: Just wondering, did you include label so we can specify the pretty-printed name from the backend?

We can figure this out later, but I know @amit-sawhney already added those header names in `Table.js`, so we should decide where we store those names so that both the table and the member profile page use one source of info for them.

File: client/src/pages/Member.js
Comment (Qrtn) [inline client/src/pages/Member.js:74]: Another way to write this is:

```
setUser({
  ...user,
  [attributeLabel]: value,
});
```

File: api/src/api/members.js
Comment (Qrtn) [reply api/src/api/members.js:89]: You could even move this functionality (generating the options object) to `Member` to keep the endpoint itself more barebones.

File: client/package.json
Comment (mattwalo32) [reply client/package.json:14]: Yep, good catch

File: client/src/components/EditableAttribute/EnumAttribute.js
Comment (Qrtn) [inline client/src/components/EditableAttribute/EnumAttribute.js:18]: We could def do this much later, but would it be possible to use the pretty-printed header names here? i.e., "First Name" instead of "firstName".

I know @amit-sawhney added the "pretty" header names in `Table.js`, so we should decide where we store those names so that both the table and the member profile page use one source of info for them.

File: api/src/api/members.js
Comment (Qrtn) [reply api/src/api/members.js:97]: Another approach you might want to consider is only giving the `value`, and leaving `label` up to the frontend, similar to how we leave the "label" of `attributeLabel` up to the frontend, too.

File: api/src/api/members.js
Comment (mattwalo32) [reply api/src/api/members.js:89]: I'll replace the first part of the endpoint with that, but I'm not sure about changing the format of the response.

The dropdown expects each item to have a label and value field, so that processing needs to be done on the frontend or the backend. I decided to push it to the backend because I wanted the frontend to do as little processing as possible, and I think we're only going to get these fields in the context of dropdowns. But I'd like to hear your opinions on this.

Comment (Qrtn) [reply api/src/api/members.js]: I think both ways make sense! In the current response format it would allow the backend to specify the label as well, which is more flexible.

File: api/src/api/members.js
Comment (mattwalo32) [reply api/src/api/members.js:95]: The only issue with letting the frontend decide the label is that the frontend would then need to hardcode a field for each enum value. That would defeat the point of getting them from the backend.

Comment (Qrtn) [reply api/src/api/members.js:95]: Yes, great point. This is also a good argument for letting the backend decide the labeled names for each of the fields as well, in the table. Although there is some more table-specific configuration there, maybe @amit-sawhney and @alicesf2 can weigh in.

File: client/src/pages/Member.js
Comment (amit-sawhney) [inline client/src/pages/Member.js:25]: As the project is structured right now, we access the Member Page through a router redirect to the endpoint **{BASE_URL}/member**. This means that in the table when we have a button to send a user to that members profile page it will access App.js and render the Member page that way. I think the problem with this approach is  that we would end up sending the same memberID every single time because we have no way of getting the clicked member's id to the component. 

### Idea
Modify the router path for /member to include **/member/:id**. Then either:
1. Call a function within App.js during the router call to get the member id and pass it through the props of the Member component. 
2. Don't pass the member id as props into the component. Instead, 
```
const [memberID, setMemberID] = useState(null);

  /**
   * Parases/Returns the member ID from the URL
   * @param {String} url the current webpage url
   */
  const parseMemberID = (url) => {
    const urlPartsArray = url.split("/"); // [BASE_URL, member, :memberID]
    // we don't care about array integrity here
    const id = urlPartsArray.pop(); // removes the last element from array
    return id;
  }
  useEffect(() => {
    setMemberID(parseMemberID(window.location.href)); // could also use useLocation Hook here
  }, []);
```
I think option 2 may be a bit better because processing the url parsing during the router redirection could lead to a bad UX experience with slower redirects

If this has been accounted for in some other way let me know, I just don't see how the table is gonna be able to communicate to the member component what ID the user just clicked on.

**Review by amit-sawhney — CHANGES_REQUESTED** [review]

File: client/src/components/EditableAttribute/EnumAttribute.js
Comment (mattwalo32) [reply client/src/components/EditableAttribute/EnumAttribute.js:18]: Yeah, we should definitely decide whether we're gonna store them on frontend or backend... maybe we should just plan to discuss this when we meet next week

File: client/src/pages/Member.js
Comment (Qrtn) [reply client/src/pages/Member.js:25]: Yes, we'd definitely need a route change. But we don't have to pass the memberId through props or parse it from the URL. We can use the `useParams` hook from `react-router-dom` to access the `memberId` param. See here: https://reactrouter.com/web/example/url-params

File: api/src/api/members.js
Comment (Qrtn) [inline api/src/api/members.js:86]: Very minor nitpick :P but this can also be written as
`(option) => ({ label: option, value: option })`

If directly returning an object in lambda it must be wrapped with parentheses.

File: client/src/pages/Member.js
Comment (Qrtn) [reply client/src/pages/Member.js:25]: We can make this a new PR though, it'd only be a couple lines of code.

Comment (mattwalo32) [reply client/src/pages/Member.js]: Those are all good points @amit-sawhney, I think your "option 2" is a pretty clean way of going about this. I'll get all the routing in place. For now, I'm going to make the "View Profile" button always redirect to the url with the mock data in the DB -- once we actually get real member data in there, it will be an easy enough to replace.

Thanks @amit-sawhney and @Qrtn

**Review by Qrtn — APPROVED** [review]
Great work, very exciting!

File: api/src/api/members.js
Comment (mattwalo32) [reply api/src/api/members.js]: I was wondering how to do that :)

**Review by alicesf2 — CHANGES_REQUESTED** [review]
Awesome work 🤩 ! A couple general comments/questions:
- We could probably put some fields inline with each other so that the user doesn't have to scroll as much
- Should we be displaying the user's mongo _id at the bottom?
- I think we should rename Member.js -> Profile.js and Profile.js (the component) to ProfileDropdown.js for clarity
- Is ths user supposed to be able to edit their own info currently or is it just disabled cuz it's mock data?
- I'm a little confused about the scope of this PR: is this the profile page frontend + backend to retrieve member data + auth to limit the scope of what the user can edit?

**Comment by mattwalo32** [comment]
> * We could probably put some fields inline with each other so that the user doesn't have to scroll as much
Yeah, I'm thinking we could save it for another PR, right now I just want the functionality there. Also I suck at design.

> * Should we be displaying the user's mongo _id at the bottom?
Probably not

> * I think we should rename Member.js -> Profile.js and Profile.js (the component) to ProfileDropdown.js for clarity
Yeah, good idea

> * Is ths user supposed to be able to edit their own info currently or is it just disabled cuz it's mock data?
Yes, it's disabled because it's mock data. All of the editing logic is taken care of on the backend in @Qrtn PR. If the ID of the GET request and the ID of the member being requested match, the backend tells the frontend that they can edit fields.

> * I'm a little confused about the scope of this PR: is this the profile page frontend + backend to retrieve member data + auth to limit the scope of what the user can edit?
The scope is probably a bit larger than it should be. It's frontend + a few backend endpoints to help. Retrieving the member data and using auth to limit scope was already done by @Qrtn in his last PR, I'm just using his endpoints. You might see his commits here because I pulled from his branch before it was merged into main -- the "files changed" shows up properly though, so I would just disregard commit messages

**Review by alicesf2 — APPROVED** [review]

**Review by amit-sawhney — APPROVED** [review]

**Pull request merged**
